### PR TITLE
[Do Not Merge] Add Microsoft.Security Insights APIs

### PR DESCRIPTION
**Pull request by ranisha2**
Add Insights APIs:

GET Insights for a specific resource
List Insights for a specific resource
APIs are not public yet. We need to review this and merge only after exposing the APIs in production.

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by markcowl** [comment]
@ranisha2 You will need to fix the violations above see the how to fix it link (https://aka.ms/ci-fix) for details.

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (guywaldman) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:76]: Do we want "Insights" (upper I) or "insights"?

You used `Microsoft.Security/insights` for the type name. I think we should go with lowercase here, no?

**Review by guywaldman — APPROVED** [review]

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:76]: Right, will update.

**Comment by ranisha2** [comment]
@markcowl Can you please take a look?

**Comment by ranisha2** [comment]
@markcowl A friendly reminder to review

**Comment by markcowl** [comment]
@ranisha2 stillo need to fix the linting violations above and fix the merge conflicts

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (markcowl) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:167]: why is this example inline in a swagger spec?

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (markcowl) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:36]: Are there any operations to create these?  Are there any scopes other than a resource?  (subscription? resourceGroup?)

**Review by markcowl — CHANGES_REQUESTED** [review]

**Comment by markcowl** [comment]
@ranisha2 also, please fill out the review checklist - once this is done, an ARM review will automaticlaly be requested.

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (varunkumta) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:156]: isn't this the same as the resource id in the url? if so, why is it needed?

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (varunkumta) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:167]: please specify only the schema in this file

**Review by varunkumta — CHANGES_REQUESTED** [review]
please take a look at review comments

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:36]: There's no operations to create this.
For now we intend to support insights on resources only

**Comment by ranisha2** [comment]
> @ranisha2 stillo need to fix the linting violations above and fix the merge conflicts

@markcowl The failing check is  R3023 - OperationsAPIImplementation, and I believe that this is not required to pass.
Please correct me if I'm wrong (Operations API is not related to this)

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:156]: It's the same. The resourceId is part of the URI.
We might want to support a List operation later, so we considered adding it here.
It still can be extracted from the resourceId, but I don't see any harm of keeping it here

**Comment by markcowl** [comment]
@ranisha2 Operatiosn API is required - you can follow http://aka.ms/ci-fix if you think you are eligible for an exception.  
Also, you will need to address the braking changes reported by the python SDK automation check:
```
Model Device no longer has parameter sensor_name
[ChangeLog]   - Model Device no longer has parameter zone_name
[ChangeLog]   - Model Device no longer has parameter site_name
```
You still need an ARM signoff - if you have addressed all of the ARM feedback, let me know, and I will chaneg the label so they can sign off.

**Comment by ranisha2** [comment]
> @ranisha2 Operatiosn API is required - you can follow http://aka.ms/ci-fix if you think you are eligible for an exception.
> Also, you will need to address the braking changes reported by the python SDK automation check:
> 
> ```
> Model Device no longer has parameter sensor_name
> [ChangeLog]   - Model Device no longer has parameter zone_name
> [ChangeLog]   - Model Device no longer has parameter site_name
> ```
> 
> You still need an ARM signoff - if you have addressed all of the ARM feedback, let me know, and I will chaneg the label so they can sign off.

@markcowl 
Operations API is implemented:
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/operations.json

The check is failing on a different API:
Operations API must be implemented for '/providers/Microsoft.Security/operations'.
New: Microsoft.Security/preview/2020-08-06-preview/iotRecommendationTypes.json#L36

Pyhton SDK checks failures are fixed (After syncing)

**Comment by markcowl** [comment]
@ranisha2 You will need to fix the merge conflicts so that ci validations can run.  Also, marked for ARM re-review based on your comments

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (RamyasreeChakka) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:156]: @ranisha2 It's not the common pattern followed in ARM. I would also recommend to remove this redundant property. As this property is not readOnly and configurable by user, you will need to do extra checks like 'resourceId' in the URL should be same as that of the value provided in resource payload.

Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:156]: @RamyasreeChakka Actually it's a read only property. I will set it as a read only.

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ravbhatnagar) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:103]: Please keep in mind that ARM has a size limit on the response payload. Based on the current modeling, if the response size continues to grow, it will problematic. I see there are things like catalogs and labels. not sure how these are created, but can this result in a huge payload for a certain scope? If yes, may be you can create child resource types for those under insights. The insights can return some common properties and rest is via the child type which can be paginated etc since it will become a resource. Does this usecase make sense? please think about it if it can be valid.
Do the customers get insight over a resource like VM or DB or can they get it for the subscription?

**Review by ravbhatnagar — COMMENTED** [review]
@ranisha2 - couple of comments.

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:76]: Hi @ravbhatnagar 
Insight name is set as an enum (See InsightName parameter below)

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (ranisha2) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:103]: @ravbhatnagar The response is aggregative and the number of classifications/labels is relatively small.
We don't expect to reach large payloads that will exceed the size limit that ARM has (Not even close to the limit).
We'll keep monitoring this and make sure that this does not happen.

**Comment by ravbhatnagar** [comment]
SIgning off from ARM side.

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (markcowl) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:168]: This modeling makes it impossible for many clients to deserialize and work with this data.  Is this a polymorphic type with different possible formats?  Is there a set of known properties and then an additional property bag which is designed to be unstructured?  From the look of the examples, these appear to be fairly structured.

**Review by markcowl — CHANGES_REQUESTED** [review]

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json
Comment (guywaldman) [reply specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/insights.json:168]: Hi Mark, this is a polymorphic type - though we currently support only one type of insight (as indicated in the example), we plan to follow this up with other types in the future, so this is by design.

**Review by markcowl — APPROVED** [review]

**Review by markcowl — CHANGES_REQUESTED** [review]
This PR cannot be merged until the linting issue is resolved.  Since this has been signed off by ARM, you should put in a suppression for the linting issue:https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process 

Again, since it has been signed off, you can simply suppress this rule in your readme as outlined in the doc.

**Comment by markcowl** [comment]
@ranisha2 Do you still want to merge this PR?  I will close at EOW (1/8/21) if there is no activity.

**Comment by ranisha2** [comment]
markcowl  we want to merge it once we are ready. please keep it open for now.
I will update the examples soon, and will let you know once it's ready to be merged.

**Comment by ranisha2** [comment]
@markcowl The examples were updated, but we're still not ready to merge the PR.
Can you add the DoNotMerge label please?

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/examples/Insights/GetStorageClassificationInsights_example.json
Comment (eliagrady) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/examples/Insights/GetStorageClassificationInsights_example.json:3]: missing prefix slash before 'subscriptions'

File: specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/examples/Insights/ListStorageInsights_example.json
Comment (eliagrady) [inline specification/security/resource-manager/Microsoft.Security/preview/2020-08-06-preview/examples/Insights/ListStorageInsights_example.json:3]: same

**Review by eliagrady — CHANGES_REQUESTED** [review]
minor examples fixes

**Comment by markcowl** [comment]
@ranisha2 please let me know when we are ready to take action on this PR

**Comment by markcowl** [comment]
@ranisha2  wil close this at the end of the week if there is no activity - we can always re-open if needed,

**Comment by markcowl** [comment]
@ranisha2 Closing for lack of activity.  Please let us know if we need to reopen.

**Pull request closed without merging**
